doc/man/gnunet-pseudonym.1
.TH GNUNET-PSEUDONYM "1" "23 Dec 2006" "GNUnet"
.SH NAME
gnunet\-pseudonym \- create, delete or list pseudonyms
.SH SYNOPSIS
.B gnunet\-pseudonym
[options]
.SH DESCRIPTION
.PP
gnunet\-pseudonym is a tool for managing pseudonyms and namespaces.  A pseudonym is the persona that controls a namespace.  As such, it is identical to a public\-private RSA key pair.  A namespace is a collection of files that have been signed by the corresponding private RSA key.  A namespace is typically associated with a nickname and other metadata which is kept in a specially named file in the namespace.

Namespaces are an important tool for providing assurances about content integrity and authenticity in GNUnet.  Since all of the content in the namespace must have been provided by the same entity, users can form an opinion about that entity and learn to search (or avoid) certain namespaces.

gnunet\-pseudonym can be used to list all of the pseudonyms that were created locally, to create new pseudonyms, to delete existing pseudonyms (the namespace will continue to exist, but it will be impossible to add additional data to it) and to list all of the namespaces (with their meta-data) known to the local user.  By default, gnunet\-pseudonym lists all pseudonyms created locally and all of the namespaces that were discovered so far.

Creating a new pseudonym requires using the \-C option together with a nickname that is to be used for the namespace.  Nicknames must be unique for each user, global uniqueness is desireable but not necessary.  If two namespaces in GNUnet use the same nickname all GNUnet tools will display the nickname together with the unique namespace identifier (which is derived from the public key and hence guaranteed to be unique) to avoid ambiguity.  Additional options can be passed together with the \-C option to provide additional meta-data that describes the namespace.  Possible meta-data includes the 'realname' of the person controlling the namespace, a description, the mime-type for content in the namespace (useful if the namespace is dedicated to some specific type of content) and contact information.  One important piece of meta-data that can be specified is the identifier of a document root, that is the name of a file in the namespace that is a portal to the rest of the content.  This is useful to help users find this root in the absence of conventions.  Note that all of this meta-data is optional and should never be trusted blindly.

gnunet\-pseudonym also lists the meta-data available for other namespaces.  Namespaces can be discovered whenever the peer obtains the namespace advertisement that is created at the time where the pseudonym is created.  Namespace advertisements can be found in directories (not implemented), ordinary keyword-based searches (by default gnunet\-pseudonym publishes the namespace advertisement under the keyword 'namespace', but the \-k option can be used to specify other keywords) and under the 'all-zeros' identifier of the respective namespace (using a namespace-search if the namespace ID is already known).

For more details about GNUnet namespaces and content encoding please read the 'Encoding for Censorship-resistant Sharing' (ECRS) paper which can be found on the GNUnet webpage.

.TP
\fB\-a \fILEVEL\fR, \fB\-\-anonymity=LEVEL\fR
set desired level of sender anonymity.  Default is 1.
.TP
\fB\-A\fR, \fB\-\-automate\fR
Start a (new) collection.  Works only in conjunction with the \-C option.  A collection is an automatically managed directory in a namespace.  In essence, after starting the collection every file that you insert into GNUnet will automatically be placed into the collection.  Other users can browse your collection and be certain (thanks to cryptography) that all of these files were inserted into GNUnet by the same user (they do not necessarily know who it is, but if you specify your realname (\-r) they will be able to see that handle).  Collections are useful for establishing a reputation for your GNUnet content, such that readers can form an opinion about quality and availability.  Namespaces can be used to achieve the same thing, but collections are automatic and thus less work for you.

Using collections has some security implications since it is possible for an adversary to see that all of these files originate from the same user.  This may help a correlation attack to break anonymity.  Nevertheless we encourage using collections, they are likely to be the right choice for most users.
.TP
\fB\-C NAME\fR, \fB\-\-create=NAME\fR
Creates a new pseudonym with the given NAME.
.TP
\fB\-D NAME\fR, \fB\-\-delete=NAME\fR
Delete the pseudonym with the given NAME.
.TP
\fB\-e EMAIL\fR, \fB\-\-email=EMAIL\fR
Include EMAIL a contact address to contact the author of the namespace (use with \-C).
.TP
\fB\-E\fR, \fB\-\-end\fR
End a collection.  This option is the opposite of the \-a option in that it stops the collection.  Note that currently, once you stop a collection you can never restart it.  However, you can start a new collection.  There can only be one collection at any given point in time for a particular user.
.TP
\fB\-h\fR, \fB\-\-help\fR
print help page
.TP
\fB\-k KEYWORD\fR, \fB\-\-keyword=KEYWORD\fR
Publish the namespace advertisement under the keyword 'KEYWORD'.  Default is 'namespace' (use with \-C).  You can specify \-k multiple times.  In that case, the namespace will be published under each of those keywords.
.TP
\fB\-l\fR, \fB\-\-local\-only\fR
only display local namespaces (those that we can extend with content because we created them)
.TP
\fB\-m MIMETYPE\fR, \fB\-\-mimetype=MIMETYPE\fR
Advertise that the namespace contains files of the given MIMETYPE (use with \-C).
.TP
\fB\-n\fR, \fB\-\-no\-advertisement\fR
Do not generate an advertisement for the namespace (use with \-C).
.TP
\fB\-p \fIPRIORITY\fR, \fB\-\-prio=\fIPRIORITY\fR
Set the priority of the inserted content (default: 365).  If the local database is full, GNUnet will discard the content with the lowest ranking.  Note that ranks change over time depending on popularity.  The default should be high enough to preserve the locally inserted content in favor of content that migrates from other peers.
.TP
\fB\-q\fR, \fB\-\-quiet\fR
Do not print the list of pseudonyms (only perform create or delete operation).
.TP
\fB\-r NAME\fR, \fB\-\-realname=NAME\fR
Claim that the name of the author of the content in the namespace in 'real' life is NAME (use with \-C).
.TP
\fB\-R IDENTIFIER\fR, \fB\-\-root=IDENTIFIER\fR
Specify the identifier for the root of the namespace.  Used in the namespace advertisement to tell users that find the namespace advertisement about an entry-point into the namespace (use with \-C).
.TP
\fB\-s ID:VALUE\fR, \fB\-\-set-rating=ID:VALUE\fR
Change the rating for the namespace identified by ID by VALUE.  For example, "\-s test:-3" decrements the rating of the pseudonym "test" by 3.  Note that ratings are purely local.  Each user has his own independent rating of namespaces.  The rating is merely a way for each user to keep track of his own experience with a given namespace.
.TP
\fB\-u URI\fR, \fB\-\-uri=URI\fR
Include URI as an address where additional information about the namespace can be found (use with \-C).

.SH FILES
.TP
~/.gnunet/data/pseudonyms/
Directory where the pseudonyms are stored
.TP
~/.gnunet/state.sdb/collection
File where information about the currently active collection is kept (if any)
.SH "REPORTING BUGS"
Report bugs by using mantis <https://gnunet.org/bugs/> or by sending electronic mail to <gnunet\-developers@gnu.org>
.SH "SEE ALSO"
\fBgnunet\-insert\fP(1), \fBgnunet\-search\fP(1), \fBgnunetd\fP(1)


ChangeLog
Mon Nov 16 09:25:47 CET 2009
	Allow passing of command-line arguments using GNUNET_ARGS
	environment variable (see Mantis #1467).

Mon Apr 27 22:18:19 CEST 2009
	Monitor low-traffic connections over a longer period
	of time before terminating them.

Fri Feb 27 22:05:50 MST 2009
	Releasing 0.8.0c.

Sun Feb  1 19:51:40 MST 2009
	Fixed Mantis #1429 (struct padding could cause sizeof
	to be different on AMD64, using gcc-ism "packed" to
	avoid this problem).

Sun Feb  1 00:37:16 MST 2009
	Fixed problem in DHT client protocol that could cause
	PUT requests from clients to be (frequently) lost.

Sun Nov  2 16:06:47 MST 2008
	Fixed problem with MySQL disconnects in one module
	impacting MySQL correctness in another module (caused
	crashes).

Sat Aug 16 00:18:34 CEST 2008
	Improved keyword extraction
	Fixed expiration time of keywords
	Releasing 0.8.0b.

Sun Jul 20 15:27:11 MDT 2008
	Releasing 0.8.0a.

Thu Jul 10 00:59:31 MDT 2008
	Fixed bug where datastore resizing would not 
	result in bloomfilter update during gnunet-update.

Sat Jun 28 23:43:00 MDT 2008
	Fixed bug preventing namespace advertisements from
	being processed.

Sat Jun 14 00:54:39 MDT 2008
	Releasing 0.8.0.

Fri Jun 13 23:51:00 MDT 2008
	GNUnet now compiles on OpenBSD.

Wed Jun 11 22:17:09 MDT 2008
	Made state of gnunet-auto-share persistent. 
	Added support for multiple directories.  Improved
	FSUI/ECRS APIs to reduce the number of threads needed
	for probing.  Increased number of concurrent probes
	allowed.

Sat Jun  7 17:54:49 MDT 2008
	Simplified namespaces a lot.  Shorter timeouts for
	trying to connect to gnunetd to prevent interactive
	apps from blocking too long when gnunetd is not
	running.

Sun Jun  1 21:11:54 MDT 2008
	Releasing 0.8.0pre1.

Sun Jun  1 13:41:27 MDT 2008
	Split of libgnunetpseudonym from libgnunetnamespace
	(with major API extension and changes).

Tue May  6 04:10:02 MDT 2008
	Added simple sanity check for [MODULES] values.
	Added "-s" and "-q" options to gnunet-peer-info.

Thu Apr 24 00:05:17 MDT 2008
	Simplified FSUI stop/abort/pause/resume API calls
	(no need to pass FSUI context anymore).

Mon Apr 21 21:05:20 MDT 2008
	Adding buffered IO for FSUI's serialize and
	deserialize code.

Sun Apr 20 20:01:20 MDT 2008
	Fixed bug in HTTP transport causing lonely messages
	(responses to GET) not to be delivered.

Mon Mar 24 21:15:36 MDT 2008
	Fixed various problems with downloading locally
	indexed large files (downloading large files from
	other peers should not have been a problem).
	
Sun Mar  2 15:33:41 MST 2008
	Added option to allow user to disable IPv6 support.
	Releasing 0.8.0pre0.

Sat Mar  1 21:14:02 MST 2008
	Fixed some dstore performance issues.
	Made transports alternate between trying IPv4 and IPv6
	if both are available (since one of them maybe
	misconfigured).

Mon Feb 25 23:25:48 MST 2008
	Cleanup of the DNS code: removed GNUNET_IPvXAddress
	and GNUNET_PRIP; centralized all DNS lookup code in 
	util/network/dns.c; improved IPv6 support for 
        network-client code.

Mon Feb 25 00:01:27 MST 2008
	Added asynchronous search and download methods for 
	ECRS library.  FSUI now can do with only one thread
	per search or download (until now, we had two 
	threads per search / download).

Tue Feb 19 20:35:28 MST 2008
	Updated database schemata to support O(1) operations
	even if there are N files under the same keywords.
	This breaks database backwards compatibility (and
	no migration code is provided).

Mon Feb 18 19:47:37 MST 2008
	Added new options -u and -s to gnunet-insert.

Sat Feb 16 21:37:33 MST 2008
	Implemented gnunet-auto-share for automatic
	sharing of directories.

Thu Feb 14 01:02:32 MST 2008
	Made HTTP transport work over IPv6 as well.

Wed Feb 13 22:59:07 MST 2008
	Modified mySQL and SQLite datastores to return diverse
	result sets during (partial) hash-based get iterations
	(should result in more diverse keyword result sets).

Tue Feb 12 23:54:34 MST 2008
	First shot at integrated hostlist HTTP server based
	on libmicrohttpd.

Sat Feb  9 22:06:23 MST 2008
	Minor changes to the ECRS/FSUI APIs for searches
	(no more timeouts and/or max-results).  Clients who
	need these features could easily implement them 
	directly (and they are not really useful to begin
	with).  
	Also, the IPv4 and IPv6 protocols are being integrated
	so that we only listen to one port (running both IPv4
	and IPv6 over the same port).  This break transport
	compatibility, but given that we're breaking protocol
	compatibility for file-sharing already, this may 
	actually be a bit of a good thing (since it will now
	be possible to tell if peers running the new GAP code
	are connected to other peers running the new GAP code).
	
Sun Feb  3 13:17:09 MST 2008
	Dramatic changes to the GAP implementation (breaking
	protocol compatibility).  Essentially, we can save
	a few bytes in each reply.  More importantly, the new
	code allows the searching client to specify a set of
	replies that are not desired (hopefully helping to
	dramatically increase the diversity of search replies
	obtained over time).  Note that the actual encoding
	and databases are not affected (just P2P protocol).
	The update is not complete yet, but should compile.

Tue Jan  8 20:07:20 MST 2008
	Added option for testing ("make check") to use weak(er)
	PRNG for key generation (thanks to Werner Koch for 
	pointing out how to do it).

Thu Dec 22 20:10:37 MST 2007
	Releasing GNUnet 0.7.3.

Sun Dec  9 14:34:32 MST 2007
	Implemented MySQL version of the dstore-module.  This
	means that sqLite is now again truly optional.

Sat Dec  8 15:15:53 MST 2007
	Integrated F2F topology into main topology module 
	(options are used to choose between F2F-only (old F2F), 
	 minimum number of friend connections (NEW) or entirely
	 arbitrary connection set (default)).

Thu Dec  6 20:51:17 MST 2007
	The SMTP transport is working again.

Thu Nov 22 20:49:16 MST 2007
	United all libgnunetutil_*.so libraries into one big
	libgnunetutil.so library (eliminates issues for binary

src/applications/fs/tools/gnunet-pseudonym.c
/*
     This file is part of GNUnet.
     (C) 2002, 2003, 2004, 2005, 2006, 2008, 2009 Christian Grothoff (and other contributing authors)

     GNUnet is free software; you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published
     by the Free Software Foundation; either version 2, or (at your
     option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     General Public License for more details.

     You should have received a copy of the GNU General Public License
     along with GNUnet; see the file COPYING.  If not, write to the
     Free Software Foundation, Inc., 59 Temple Place - Suite 330,
     Boston, MA 02111-1307, USA.
*/

/**
 * @file applications/fs/tools/gnunet-pseudonym.c
 * @brief create, list or delete pseudoynms
 * @author Christian Grothoff
 */

#include "platform.h"
#include "gnunet_directories.h"
#include "gnunet_collection_lib.h"
#include "gnunet_namespace_lib.h"
#include "gnunet_util.h"

static struct GNUNET_GE_Context *ectx;

static struct GNUNET_GC_Configuration *cfg;

static int start_collection;

static int stop_collection;

static int be_quiet;

static int no_advertisement;

static char *delete_name;

static char *create_name;

static char *set_rating;

static char *root_name;

static int local_only;

static unsigned int anonymity;

static unsigned int priority = 365;

static GNUNET_CronTime expiration = 2 * GNUNET_CRON_YEARS;

static char *cfgFilename = GNUNET_DEFAULT_CLIENT_CONFIG_FILE;

static struct GNUNET_MetaData *meta;

static struct GNUNET_ECRS_URI *advertisement;

/**
 * All gnunet-pseudonym command line options
 */
static struct GNUNET_CommandLineOption gnunetpseudonymOptions[] = {
  {'a', "anonymity", "LEVEL",
   gettext_noop ("set the desired LEVEL of sender-anonymity"),
   1, &GNUNET_getopt_configure_set_uint, &anonymity},
  {'A', "automate", NULL,
   gettext_noop ("automate creation of a namespace by starting a collection"),
   0, &GNUNET_getopt_configure_set_one, &start_collection},
  GNUNET_COMMAND_LINE_OPTION_CFG_FILE (&cfgFilename),   /* -c */
  {'C', "create", "NICKNAME",
   gettext_noop ("create a new pseudonym under the given NICKNAME"),
   1, &GNUNET_getopt_configure_set_string, &create_name},
  {'D', "delete", "NICKNAME",
   gettext_noop ("delete the pseudonym with the given NICKNAME"),
   1, &GNUNET_getopt_configure_set_string, &delete_name},
  {'E', "end", NULL,
   gettext_noop ("end automated building of a namespace (ends collection)"),
   0, &GNUNET_getopt_configure_set_one, &stop_collection},
  GNUNET_COMMAND_LINE_OPTION_HELP (gettext_noop ("Create new pseudonyms, delete pseudonyms or list existing pseudonyms.")),     /* -h */
  GNUNET_COMMAND_LINE_OPTION_LOGGING,   /* -L */
  {'l', "local-only", NULL,
   gettext_noop
   ("only display local namespaces"),
   0, &GNUNET_getopt_configure_set_one, &local_only},
  {'k', "keyword", "KEYWORD",
   gettext_noop
   ("use the given keyword to advertise the namespace (use when creating a new pseudonym)"),
   1, &GNUNET_ECRS_getopt_configure_set_keywords, &advertisement},
  {'m', "meta", "TYPE:VALUE",
   gettext_noop ("specify metadata describing the namespace or collection"),
   1, &GNUNET_ECRS_getopt_configure_set_metadata, &meta},
  {'n', "no-advertisement", NULL,
   gettext_noop
   ("do not generate an advertisement for this namespace (use when creating a new pseudonym)"),
   0, &GNUNET_getopt_configure_set_one, &no_advertisement},
  {'p', "priority", "PRIORITY",
   gettext_noop ("specify the priority of the content"),
   1, &GNUNET_getopt_configure_set_uint, &priority},
  {'q', "quiet", NULL,
   gettext_noop ("do not list the pseudonyms from the pseudonym database"),
   0, &GNUNET_getopt_configure_set_one, &be_quiet},
  {'R', "root", "IDENTIFIER",
   gettext_noop
   ("specify IDENTIFIER to be the address of the entrypoint to content in the namespace (use when creating a new pseudonym)"),
   1, &GNUNET_getopt_configure_set_string, &root_name},
  {'s', "set-rating", "ID:VALUE",
   gettext_noop ("set the rating of a namespace"),
   0, &GNUNET_getopt_configure_set_string, &set_rating},
  GNUNET_COMMAND_LINE_OPTION_VERSION (PACKAGE_VERSION), /* -v */
  GNUNET_COMMAND_LINE_OPTION_VERBOSE,
  GNUNET_COMMAND_LINE_OPTION_END,
};

static int
itemPrinter (EXTRACTOR_KeywordType type, const char *data, void *closure)
{
  printf ("\t%20s: %s\n",
          dgettext (LIBEXTRACTOR_GETTEXT_DOMAIN,
                    EXTRACTOR_getKeywordTypeAsString (type)), data);
  return GNUNET_OK;
}

static void
printMeta (const struct GNUNET_MetaData *m)
{
  GNUNET_meta_data_get_contents (m, &itemPrinter, NULL);
}

static int
namespacePrinter (void *unused,
                  const GNUNET_HashCode * id,
                  const struct GNUNET_MetaData *md, int rating)
{
  GNUNET_EncName enc;
  GNUNET_HashCode nsid;
  int cpos;
  char *namespaceName;

  if ( (local_only != 0) &&
       (GNUNET_OK != GNUNET_ECRS_namespace_test_exists (NULL,
							cfg,
							id)) )
    return GNUNET_OK;
  namespaceName = GNUNET_pseudonym_id_to_name (ectx, cfg, id);
  GNUNET_hash_to_enc (id, &enc);
  if ( (namespaceName != NULL) &&
       (0 == strcmp (namespaceName, (char *) &enc)) )
    printf (_("Namespace `%s' has rating %d.\n"), namespaceName, rating);
  else
    printf (_("Namespace `%s' (%s) has rating %d.\n"),
            namespaceName != NULL ? namespaceName : "<unknown>",
	    (char *) &enc, rating);
  printMeta (md);
  if (set_rating != NULL)
    {
      int delta;
      char *set;

      set = set_rating;
      delta = 0;
      cpos = 0;
      while ((set[cpos] != '\0') && (set[cpos] != ':'))
        cpos++;
      if (((((strlen ((char *) &enc) + 1 == cpos) &&
             (0 == strncmp (set,
                            (char *) &enc,
                            cpos)))) ||
           ((namespaceName != NULL) &&
            (strlen (namespaceName) == cpos) &&
            (0 == strncmp (set, namespaceName, cpos)))) && (set[cpos] == ':'))
        {
          delta = strtol (&set[cpos + 1], NULL, /* no error handling yet */
                          10);
        }

      if (delta != 0)
        {
	  rating = GNUNET_pseudonym_rank (ectx, cfg, &nsid, delta);
	  printf (_("\tRating (after update): %d\n"), rating);
        }
    }
  printf ("\n");
  GNUNET_free_non_null (namespaceName);
  return GNUNET_OK;
}

int
main (int argc, char *const *argv)
{
  int cnt;
  int success;
  int i;
  GNUNET_HashCode nsid;
  char *rootEntry;
  struct GNUNET_ECRS_URI *rootURI;
  char *root;
  char *ns_name;

  meta = GNUNET_meta_data_create ();
  i = GNUNET_init (argc,
                   argv,
                   "gnunet-pseudonym [OPTIONS]",
                   &cfgFilename, gnunetpseudonymOptions, &ectx, &cfg);
  if (i == -1)
    {
      GNUNET_meta_data_destroy (meta);
      GNUNET_fini (ectx, cfg);
      return -1;
    }
  success = 0;                  /* no errors */
  GNUNET_CO_init (ectx, cfg);

  /* stop collections */
  if (stop_collection && (!start_collection))
    {
      if (GNUNET_OK == GNUNET_CO_collection_stop ())
        printf (_("Collection stopped.\n"));
      else
        printf (_("Failed to stop collection (not active?).\n"));
      be_quiet = 1;
    }

  /* delete pseudonyms */
  if (delete_name != NULL)
    {
      if (GNUNET_OK ==
          GNUNET_pseudonym_name_to_id (ectx, cfg, delete_name, &nsid))
        {
          if (GNUNET_OK == GNUNET_NS_namespace_delete (ectx, cfg, &nsid))
            {
              printf (_("Pseudonym `%s' deleted.\n"), delete_name);
            }
          else
            {
              success += 2;
              printf (_("Error deleting pseudonym `%s' (does not exist?).\n"),
                      delete_name);
            }
        }
      else
        {
          printf (_("\tUnknown namespace `%s'\n"), delete_name);
        }
      GNUNET_free (delete_name);
      be_quiet = 1;
    }

  /* create collections / namespace */
  if (create_name != NULL)
    {
      GNUNET_meta_data_insert (meta, EXTRACTOR_TITLE, create_name);
      if (start_collection)
        {
          /* FIXME: allow other update policies */
          if (GNUNET_OK == GNUNET_CO_collection_start (anonymity,
                                                       priority, meta))
            {
              printf ("%s", _("Started collection.\n"));
            }
          else
            {
              printf ("%s", _("Failed to start collection.\n"));
              success++;
            }
	  be_quiet = 1;
        }
      else
        {                       /* no collection */
          if (root_name == NULL)
            {
              rootEntry = "root";
            }
          else
            {
              rootEntry = root_name;
            }
          if (no_advertisement)
            {
              if (advertisement != NULL)
                GNUNET_ECRS_uri_destroy (advertisement);
              advertisement = NULL;
            }
          else
            {
              if (advertisement == NULL)
                advertisement =
                  GNUNET_ECRS_keyword_string_to_uri (ectx, "namespace");
            }
          rootURI = GNUNET_NS_namespace_create (ectx,
                                                cfg,
                                                anonymity,
                                                priority,
                                                expiration +
                                                GNUNET_get_time (),
                                                meta,
                                                advertisement, rootEntry);
          if (rootURI == NULL)
            {
              printf ("%s", _("Could not create namespace.\n"));
              success += 1;
            }
          else
            {
              GNUNET_ECRS_uri_get_namespace_from_sks (rootURI, &nsid);
              ns_name = GNUNET_pseudonym_id_to_name (ectx, cfg, &nsid);
              root = GNUNET_ECRS_uri_to_string (rootURI);
              printf (_("Namespace `%s' created (root: %s).\n"),
                      ns_name, root);
	      printf (_("Note that a number was appended to your selected name to ensure uniqueness on your system.\n"));
              GNUNET_free (ns_name);
              GNUNET_free (root);
              GNUNET_ECRS_uri_destroy (rootURI);
            }
	  be_quiet = 1;
          if (NULL != advertisement)
            GNUNET_ECRS_uri_destroy (advertisement);
        }
      GNUNET_meta_data_delete (meta, EXTRACTOR_TITLE, create_name);
      GNUNET_free (create_name);
      create_name = NULL;
    }
  else
    {
      if (start_collection)
	{
	  printf (_
		  ("You must specify a name for the collection (`%s' option).\n"),
		  "-C");
	  be_quiet = 1;
	}
    }
  if (0 == be_quiet)
    {
      /* print information about pseudonyms */
      cnt = GNUNET_pseudonym_list_all (ectx, cfg, &namespacePrinter, NULL);
      if (cnt == -1)
        printf (_("Could not access namespace information.\n"));
    }
  GNUNET_meta_data_destroy (meta);
  GNUNET_CO_done ();
  GNUNET_fini (ectx, cfg);
  return success;
}

/* end of gnunet-pseudonym.c */
